File: 15824.py
from sys import stdin

input = stdin.readline


def solve():
    mod = 1_000_000_007
    N = int(input())
    arr = [*map(int, input().split())]
    if N == 1: return 0
    arr.sort()
    power = [1] * N
    for i in range(1, N):
        power[i] = (2 * power[i-1]) % mod

    ans = 0
    for i in range(N // 2):
        ans = (ans + (power[tmp := N - 1 - i] - power[i]) * (arr[tmp] - arr[i])) % mod

    return ans


if __name__ == '__main__':
    print(solve())

File: 15824.js
const fs = require('fs');

function solve() {
	// BigInt 사용
	const MOD = 1_000_000_007n;
	const input = fs.readFileSync('/dev/stdin').toString().split('\n');
	let inputIdx = 0;
	const N = +input[inputIdx++];
	const arr = input[inputIdx++]
		.split(' ')
		.map((v) => BigInt(v))
		.sort((a, b) => Number(a - b));

	if (N === 1) return 0;

	const power = [];
	power.length = N;
	power[0] = 1n;
	for (let i = 1; i < N; i++) {
		power[i] = (2n * power[i - 1]) % MOD;
	}

	let tmp,
		ans = 0n;
	for (let i = 0; i < Math.floor(N / 2); i++) {
		tmp = N - 1 - i;
		ans = (ans + (arr[tmp] - arr[i]) * (power[tmp] - power[i])) % MOD;
		// JS 는 -5 % 3 === -2 이므로 음수일 경우 +MOD 해줌
		if (ans < 0) ans += MOD;
	}

	return Number(ans);
}

console.log(solve());
